core/src/main/java/org/apache/iceberg/BaseTransaction.java

```diff
@@ -421,6 +421,7 @@ private void commitSimpleTransaction() {
           .onlyRetryOn(CommitFailedException.class)
           .run(
               underlyingOps -> {
+                deletedFiles.clear();
```

Comment:
It's not safe to do this here because this will clear before the first commit. `applyUpdates` will not re-commit everything on the first commit because those are already completed. Instead I think this needs to be called within `applyUpdates` just before all operations are re-committed.

Reply:
Yes you're right, I stepped through the code a bit more, really the applyUpdates is more of a `reapplyUpdates` (in case the table state changed). `applyUpdates` is a no-op on the first attempt. If the deletedFiles set is cleared before the first commit and that commit succeeds we would never end up cleaning those files based on the cleanup done here https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/BaseTransaction.java#L450. Although I don't know why our tests wouldn't pick up on this, we have tests which validate we cleanup the files we would expect